### PR TITLE
server side support of upsert batch level enum dictionary

### PR DESCRIPTION
**Pull request by jshencode**

### CODE REVIEW

File: memstore/ingestion.go [hunk 1]
```diff
@@ -24,6 +24,14 @@ import (
 // HandleIngestion logs an upsert batch and applies it to the in-memory store.
 func (m *memStoreImpl) HandleIngestion(table string, shardID int, upsertBatch *UpsertBatch) error {
 	utils.GetReporter(table, shardID).GetCounter(utils.IngestedUpsertBatches).Inc(1)
+	schema, err := m.GetSchema(table)
+	if err != nil {
+		return err
+	}
+	if err = upsertBatch.ResolveEnumDict(table, schema, m.metaStore); err != nil {
```
Comment (voyager-dw) [inline memstore/ingestion.go]: is this in other review?

File: memstore/common/upsert_batch_builder.go [hunk 2]
```diff
@@ -40,13 +42,14 @@ const (
 )
 
 const (
-	UpsertBatchVersion uint32 = 0xFEED0001
+	UpsertBatchVersion1 uint32 = 0xFEED0001
```
[the remark this answers is not in the document]
Comment (jshencode) [reply memstore/common/upsert_batch_builder.go]: there is no version upgrade needed for this change

File: memstore/ingestion.go (hunk 1 above)
Comment (jshencode) [reply memstore/ingestion.go]: other review?

**Comment by jshencode** [comment]
> can we have some diagram on the current and new batch format?
old format

```
// The serialized buffer of the batch is in the following format:
//	[uint32] magic_number
//	[uint32] buffer_size
//
//	<begin of buffer>
//	[int32]  version_number
//	[int32]  num_of_rows
//	[uint16] num_of_columns
//	<reserve 14 bytes>
//	[uint32] arrival_time
//	[uint32] column_offset_0 ... [uint32] column_offset_x+1
//	[uint32] column_reserved_field1_0 ... [uint32] column_reserved_field1_x
//	[uint32] column_reserved_field2_0 ... [uint32] column_reserved_field2_x
//	[uint32] column_data_type_0 ... [uint32] column_data_type_x
//	[uint16] column_id_0 ... [uint16] column_id_x
//	[uint8] column_mode_0 ... [uint8] column_mode_x
//
//	(optional) [uint8] null_vector_0
//	(optional) [padding to 4 byte alignment uint32] offset_vector_0
//	[padding for 8 byte alignment] value_vector_0
//	...
//
//	[padding for 8 byte alignment]
//	<end of buffer>
```

new formt
```
// The serialized buffer of the batch is in the following format:
//	[uint32] magic_number
//	[uint32] buffer_size
//
//	<begin of buffer>
//	[int32]  version_number
//	[int32]  num_of_rows
//	[uint16] num_of_columns
//	<reserve 14 bytes>
//	[uint32] arrival_time
//	[uint32] column_offset_0 ... [uint32] column_offset_x+1
//	[uint32] enum_dict_length_0 ... [uint32] enum_dict_length_x
//	[uint32] column_reserved_field2_0 ... [uint32] column_reserved_field2_x
//	[uint32] column_data_type_0 ... [uint32] column_data_type_x
//	[uint16] column_id_0 ... [uint16] column_id_x
//	[uint8] column_mode_0 ... [uint8] column_mode_x
//
//	(optional) [uint8] null_vector_0
//      (optional) [uint8] enum_dict_vector_0
//	(optional) [padding to 4 byte alignment uint32] offset_vector_0
//	[padding for 8 byte alignment] value_vector_0
//	...
//
//	[padding for 8 byte alignment]
//	<end of buffer>
```

```
//	[uint32] column_reserved_field1_0 ... [uint32] column_reserved_field1_x
```
to
```
//	[uint32] enum_dict_length_0 ... [uint32] enum_dict_length_x
```

new
```
//      (optional) [uint8] enum_dict_vector_0
```
added behind
```
//	(optional) [uint8] null_vector_0
```

File: memstore/upsert_batch.go [hunk 3]
```diff
@@ -520,97 +573,63 @@ func readUpsertBatchNew(buffer []byte) (*UpsertBatch, error) {
...
-
-		columns[i] = &columnReader{columnID: columnID, columnMode: columnMode, columnUpdateMode: columnUpdateMode, dataType: columnType,
-			cmpFunc: common.GetCompareFunc(columnType)}
-
-		columnStartOffset, err := header.ReadColumnOffset(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read start offset for column %d", i)
+		if tableSchema.Schema.Columns[columnID].DisableAutoExpand {
+			// no enum resolution needed, in this case, enum will be part of table schema
+			// and translation will be done by ingestion client or subscriber
```
Comment (shz117) [inline memstore/upsert_batch.go]: how can subscriber do the translation? it doesn't know about each instance's local enum dict, right?

Comment (jshencode) [reply memstore/upsert_batch.go]: this is for columns that auto expand is disabled, in this situation, the enum expansion will need to be done by the user through schema change in controller, therefore, the enum list for those columns will be part of the table schema. In such case, subscriber or client will be able to translate the enum.

File: memstore/upsert_batch.go [hunk 4]
```diff
@@ -520,97 +573,63 @@ func readUpsertBatchNew(buffer []byte) (*UpsertBatch, error) {
...
-
-	columns := make([]*columnReader, batch.NumColumns)
-	for i := range columns {
-		columnType, err := header.ReadColumnType(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read type for column %d", i)
-		}
+// ResolveEnumDict resolves upsert batch level enum dict with instance level enum dictionary
+func (u *UpsertBatch) ResolveEnumDict(table string, tableSchema *TableSchema, metaStore metastore.MetaStore) error {
+	tableSchema.RLock()
```
Comment (voyager-dw) [inline memstore/upsert_batch.go]: better and utility function like GetEnumDict() and do the lock/unlock operation should be inside the object.

**Comment by voyager-dw** [comment]
looks good

File: memstore/upsert_batch.go (hunk 4 above)
Comment (jshencode) [reply memstore/upsert_batch.go]: we agreed early on the convention that the user will do lock and unlock consistently.

**Pull request merged**